helix-term/src/commands.rs

```diff
@@ -4401,36 +4405,54 @@ fn scroll_down(cx: &mut Context) {
     scroll(cx, cx.count(), Direction::Forward);
 }
 
-fn select_textobject_around(cx: &mut Context) {
-    select_textobject(cx, textobject::TextObject::Around);
+fn select_textobject_around_word(cx: &mut Context) {
+    let count = cx.count();
+    let (view, doc) = current!(cx.editor);
+    let text = doc.text().slice(..);
+    let selection = doc
+        .selection(view.id)
+        .clone()
+        .transform(|range| textobject::textobject_word(text, range, TextObject::Around, count));
+    doc.set_selection(view.id, selection);
 }
 
-fn select_textobject_inner(cx: &mut Context) {
-    select_textobject(cx, textobject::TextObject::Inside);
+fn select_textobject_inner_word(cx: &mut Context) {
+    let count = cx.count();
+    let (view, doc) = current!(cx.editor);
+    let text = doc.text().slice(..);
+    let selection = doc
+        .selection(view.id)
+        .clone()
+        .transform(|range| textobject::textobject_word(text, range, TextObject::Inside, count));
+    doc.set_selection(view.id, selection);
 }
 
-fn select_textobject(cx: &mut Context, objtype: textobject::TextObject) {
+fn select_textobject_around_paragraph(cx: &mut Context) {
     let count = cx.count();
-    cx.on_next_key(move |cx, event| {
-        if let Some(ch) = event.char() {
-            let (view, doc) = current!(cx.editor);
-            let text = doc.text().slice(..);
+    let (view, doc) = current!(cx.editor);
+    let text = doc.text().slice(..);
+    let selection = doc.selection(view.id).clone().transform(|range| {
+        textobject::textobject_paragraph(text, range, TextObject::Around, count)
+    });
+    doc.set_selection(view.id, selection);
+}
 
-            let selection = doc.selection(view.id).clone().transform(|range| {
-                match ch {
-                    'w' => textobject::textobject_word(text, range, objtype, count),
-                    // TODO: cancel new ranges if inconsistent surround matches across lines
-                    ch if !ch.is_ascii_alphanumeric() => {
-                        textobject::textobject_surround(text, range, objtype, ch, count)
-                    }
-                    _ => range,
-                }
-            });
-            doc.set_selection(view.id, selection);
-        }
-    })
+fn select_textobject_inner_paragraph(cx: &mut Context) {
+    let count = cx.count();
+    let (view, doc) = current!(cx.editor);
+    let text = doc.text().slice(..);
+    let selection = doc.selection(view.id).clone().transform(|range| {
+        textobject::textobject_paragraph(text, range, TextObject::Inside, count)
+    });
+    doc.set_selection(view.id, selection);
 }
 
+// TODO: add textobject for other types like parenthesis
+// TODO: cancel new ranges if inconsistent surround matches across lines
+// ch if !ch.is_ascii_alphanumeric() => {
+//     textobject::textobject_surround(text, range, objtype, ch, count)
+// }
```

Comment:
@sudormrfbin We may have to tweak the current keymap data format to be able to store extra hooks and process it for infobox to work, at the same time we need to think of what to put on the left side of the infobox since we usually display the original key but now it can be any key.

Reply:
I think a better way to do this is to introduce a method to dynamically update the infobox, something like so:

```rust
fn select_textobject(cx: &mut Context, objtype: textobject::TextObject) {
    let count = cx.count();

	cx.editor.set_autoinfo(hashmap!(
		"w" => "Word",
		"p" => "Paragraph",
		"(,{,\",[" => "Surround",
	));

	cx.on_next_key(move |cx, event| { /* ... */ }
}

```

Reply:
I think a better way is to store the `key` in the context.

Reply:
> I think a better way is to store the `key` in the context.

I don't quite understand... could you elaborate ? If we have something like the `set_autoinfo` method we can display arbitrary information in the keys and description. For example for surround we can show `(,{,",[` and not just a single key.

Reply:
I mean make `key` part of `Context` so that autoinfo can use the current flow with some tweak to allow a wildcard, then on keypress set the `Context` `key` so that the function can know what key is pressed.